doc/rtd/explanation/about-cloud-config.rst

```diff
@@ -0,0 +1,85 @@
+.. _about-cloud-config:
+
+About the cloud-config file
+***************************
+
+The ``#cloud-config`` file is a type of user data that cloud-init can consume
+to automatically set up various aspects of the system.
+
+It overrides all other types of user data, so if there is a conflict between
+the config you specify in other :ref:`user data formats <user_data_formats>`,
+the contents of the ``#cloud-config`` file will take precendence.
+
+Note that networks are not configurable via the ``#cloud-config`` file because
+:ref:`network configuration <network_config>` occurs before user data is
+consumed and applied.
+
+How do I create a cloud-config file?
+====================================
+
+The cloud-config file can be written in any valid YAML but the first line
+**must start** with ``#cloud-config``. This line identifies the file to
+cloud-init and ensures that it will be processed as intended.
+
+After the first line, every aspect of the system's configuration is controlled
+through specific cloud-init **modules**. Each module included in the
+``cloud-config`` file can be thought of as a section.
+
+Let us consider the example of the "keyboard" module. The module has a
+corresponding top-level key (``keyboard:``, in this case), and beneath that
+the various configuration parameters are defined:
+
+.. code-block:: yaml
+
+   keyboard:
+     layout: us
+     model: pc105
+     variant: nodeadkeys
+     options: compose:rwin
+
+A full list of modules can be found :ref:`on our modules page <modules>`. This
+list also shows the valid schema keys for every module, and YAML examples.
+
+Module ordering
+---------------
+
+The order of the different module "sections" is mostly unimportant and modules
+can be shown in any order -- except where there are dependencies. For example,
+if you want to create users and also put them in a specific group, then the
+``groups`` section must go before ``users`` so that the groups are created
+first.
```

Comment:
To demonstrate the issue:
This config:
```
#cloud-config
users:
  - name: chad
  - name: sally
    groups: sallygroup
groups:
 - sallygroup
 - mygroup: [chad, sally]
```
will be translated exactly the same as this config:
```
#cloud-config
groups:
 - sallygroup
 - mygroup: [chad, sally]
users:
  - name: chad
  - name: sally
    groups: sallygroup
```

Using the `users` key I can create the `chad` and `sally` users and even add a user to a particular group. However, regardless of how I order the cloud-config,
```
groups:
 - mygroup: [chad, sally]
```
will fail regardless of the yaml ordering if the `chad` and `sally` users didn't already exist on the system, because the code that creates the groups runs before the code that creates the users.

This is a quirk of the Users/Groups module that I don't think should be mentioned in this part of the documentation.

Reply:
Understood. Since this is the only case where ordering matters, I've taken out the discussion entirely and instead put a note to call it out specifically with a link to the module. Thanks for the clarification! (I will probably put this into the explanation stuff about creating users and groups, because I think it will help illustrate the quirk nicely - but I agree, this is not the place for it)